agent-connection/src/main/kotlin/com/luxoft/blockchainlab/corda/hyperledger/indy/AgentConnection.kt

```diff
@@ -1,39 +1,183 @@
+
+    /**
+     * Receives a credential request from the Indy Party represented by this connection
+     *
+     * @return
+     *          observable (Single<>) object emitting a CredentialRequestInfo upon subscription
+     */
+    fun receiveCredentialRequest(): Single<CredentialRequestInfo>
+
+    /**
+     * Sends a credential to Indy Party represented by this connection
+     *
+     * @param credential
+     *          {@code credential} - credential message
+     */
+    fun sendCredential(credential: CredentialInfo)
+
+    /**
+     * Receives a credential from the Indy Party represented by this connection
+     *
+     * @return
+     *          observable (Single<>) object emitting a single CredentialRequestInfo upon subscription
+     */
+    fun receiveCredential(): Single<CredentialInfo>
+
+    /**
+     * Sends a proof request message to the Indy Party represented by this connection
+     *
+     * @param request
+     *          {@code request} - proof request message
+     */
+    fun sendProofRequest(request: ProofRequest)
+
+    /**
+     * Receives a proof request from the Indy Party represented by this connection
+     *
+     * @return
+     *          observable (Single<>) object emitting a single ProofRequest upon subscription
+     */
+    fun receiveProofRequest(): Single<ProofRequest>
+
+    /**
+     * Sends a proof to the Indy Party represented by this connection
+     *
+     * @param proof
+     *          {@code proof} - proof message
+     */
+    fun sendProof(proof: ProofInfo)
+
+    /**
+     * Receives a proof from the Indy Party represented by this connection
+     *
+     * @return
+     *          observable (Single<>) object emitting a single ProofInfo upon subscription
+     */
+    fun receiveProof(): Single<ProofInfo>
+
+    /**
+     * Returns the connected Indy Party session DID
+     */
+    fun partyDID(): String
+
+    /**
+     * Returns self session DID
+     */
+    fun myDID(): String
+}
 
 class AgentConnectionException(obj: Any) :
         RuntimeException("AgentConnection exception: $obj")
 
+/**
+ * Convenience extension to subscribe with a single lambda on both onSuccess() and onError()
+ */
+fun <T> Single<T>.handle(handle: (message: T?, ex: Throwable?) -> Unit): Subscription = this.subscribe({ handle(it, null) }, { handle(null, it) })
+
+/**
+ * Represents a connection to Indy Agent
+ */
 interface AgentConnection {
-    fun connect(url: String, login: String, password: String)
-    fun disconnect()
 
-    fun generateInvite(): String
-    fun acceptInvite(invite: String)
-    fun waitForInvitedParty()
+    /**
+     * Connects to Indy Agent's connection endpoint
+     *
+     * @param url
+     *          Indy Agent's endpoint URL
+     * @param login
+     *          endpoint's login
+     * @param password
+     *          endpoints's password
+     * @return
+     *          Observable<Unit> emitting a single Unit upon a successful handshake
```

Comment:
JFYI: You can wrap types in javadocs into brackets like `[Observable] of [Unit]` to make them interactive.

Reply:
done in 4d7a39afc27ef94281219b69b90ffedfad7154a3